Feature blog for KEP-3333 Retroactive default StorageClass assignment (#37625)
kubernetes/website: content/en/blog/_posts/2022-12-08-retroactive-default-storage-class.md

---
layout: blog
title: "Kubernetes 1.26: Retroactive Default StorageClass"
date: 2022-12-08
slug: retroactive-default-storage-class
---

**Author:** Roman Bednář (Red Hat)

The v1.25 release of Kubernetes introduced an alpha feature to change how a default StorageClass was assigned to a PersistentVolumeClaim (PVC).
With the feature enabled, you no longer need to create a default StorageClass first and PVC second to assign the class. Additionally, any PVCs without a StorageClass assigned can be updated later.
This feature was graduated to beta in Kubernetes 1.26.

> See [Official Kubernetes documentation](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#retroactive-default-storageclass-assignment) for more details.

Reviewer: ```suggestion
See the section on [Retroactive default StorageClass assignment](/docs/concepts/storage/persistent-volumes/#retroactive-default-storageclass-assignment) in the official Kubernetes documentation for more details.
```

## Why did StorageClass assignment need improvements

Users might already be familiar with a similar feature that assigns default StorageClasses to **new** PVCs at the time of creation. This is currently handled by the [admission controller](/docs/reference/access-authn-authz/admission-controllers/#defaultstorageclass).

But what if there wasn't a default StorageClass defined at the time of PVC creation? 
Users would end up with a PVC that would never be assigned a class. 
As a result, no storage would be provisioned, and the PVC would be somewhat "stuck" at this point. 
Generally, two main scenarios could result in "stuck" PVCs and cause problems later down the road. 
Let's take a closer look at each of them.

> ### Changing default StorageClass did not have a good solution

Reviewer: Since we're talking about the two scenarios that could cause the *stuck* PVC situation, amending our subheading to reflect the below would be more appropriate.
```suggestion
### Changing default StorageClass
```

With the alpha feature enabled, there were two options admins had when they wanted to change the default StorageClass:

1. Creating a new StorageClass as default before removing the old one associated with the PVC. 
This would result in having two defaults for a short period.
At this point, if a user were to create a PersistentVolumeClaim with storageClassName set to <code>null</code> (implying default StorageClass), the newest default StorageClass would be chosen and assigned to this PVC.

2. Removing the old default first and creating a new default StorageClass.
This would result in having no default for a short time.
Subsequently, if a user were to create a PersistentVolumeClaim with storageClassName set to <code>null</code> (implying default StorageClass), the PVC would be in <code>Pending</code> state forever.
The user would have to fix this by deleting the PVC and recreating it once the default StorageClass was available.

### Resource ordering during cluster installation

> If a cluster installation tool needed to create resources that required storage, e.g., image registry, it was difficult to get the ordering right. 

Reviewer: ```suggestion
If a cluster installation tool needed to create resources that required storage, for example, an image registry, it was difficult to get the ordering right. 
```

This is because any Pods that required storage would rely on the presence of a default StorageClass and would fail to be created if it wasn't defined.

## What changed

We've changed the PersistentVolume (PV) controller to assign a default StorageClass to any unbound PersistentVolumeClaim that has the storageClassName set to <code>null</code>.

> We've also modified the PersistentVolumeClaim admission within the API server to allow the change of values from unset value to an actual StorageClass name.

Reviewer: ```suggestion
We've also modified the PersistentVolumeClaim admission within the API server to allow the change of values from an unset value to an actual StorageClass name.
```
